File: csharp/components/Strata.Diagnostic/Core/DiagnosticSuite.cs
//  ##########################################################################
//  # File Name: DiagnosticSuite.cs
//  # Copyright: 2011-2018, Sapientia Systems, LLC.
//  ##########################################################################
 
using System;
using System.Collections.Generic;

namespace Strata.Diagnostic.Core
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Represents a suite of <c>IDiagnostic</c>s that are meant to be 
    /// run as a whole or as part of a larger suite. <c>DiagnosticSuite</c> 
    /// is a <a href=
    /// "http://en.wikipedia.org/wiki/Composite_pattern">Composite</a> of 
    /// <c>IDiagnostic</c>s.
    /// </summary>
    /// 
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public 
    class DiagnosticSuite:
        AbstractDiagnostic
    {
        private readonly IDictionary<String,IDiagnostic> diagnostics;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>DiagnosticSuite</c> instance.
        /// </summary>
        /// 
        public 
        DiagnosticSuite(String name):
            base( name )
        {
            diagnostics = new Dictionary<String,IDiagnostic>();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// 
        /// </summary>
        /// 
	    public override void 
	    RunDiagnostic(IDiagnosticResult result)
	    {
		    try
		    {
			    result.BeginDiagnostic( this );
			    BeginDiagnosticMode();
			
			    foreach (IDiagnostic d in diagnostics.Values)
				    d.RunDiagnostic( result );
		    }
		    catch (DiagnosticAbortedException ae)
		    {
			    result.ReportBeginFailure( this,ae );
		    }
		    finally
		    {
			    EndDiagnosticMode();
			    result.EndDiagnostic( this );
		    }
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Adds a <c>IDiagnostic</c> to 
        /// the <c>DiagnosticSuite</c>.
        /// </summary>
        /// 
        /// <param name="d">diagnostic being added to suite</param>
        /// 
        /// <precondition>
        /// this.HasDiagnostic( d.Name ) == false
        /// </precondition>
        /// <postcondition>
        /// this.HasDiagnostic( d.Name ) == true
        /// </postcondition>
        /// <postcondition>
        /// this.GetDiagnostic( d.Name ) == d
        /// </postcondition>
        /// 
	    public void 
	    AddDiagnostic(IDiagnostic d)
	    {
		    diagnostics.Add( d.Name,d );
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Removes the <c>IDiagnostic</c> with the specified 
        /// name from the <c>DiagnosticSuite</c>.
        /// </summary>
        /// 
        /// <param name="name">
        /// Name of diagnostic begin removed from suite
        /// </param>
        /// 
        /// <precondition>
        /// this.HasDiagnostic( name ) == true
        /// </precondition>
        /// <postcondition>
        /// this.HasDiagnostic( name ) == false
        /// </postcondition>
        /// 
	    public void 
	    RemoveDiagnostic(String name)
	    {
		    diagnostics.Remove( name );
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Returns the <c>IDiagnostic</c> associated 
        /// with the specified name.
        /// </summary>
        /// 
        /// <param name="name">Name of the diagnostic being queried</param>
        /// <returns>diagnostic associated with name</returns>
        /// 
	    public IDiagnostic 
	    GetDiagnostic(String name)
	    {
		    return diagnostics[name];
	    }
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Queries the <c>DiagnosticSuite</c> if it contains an
        /// <c>IDiagnostic</c> with the specified name.
        /// </summary>
        /// 
        /// <param name="name">Name of the diagnostic being queried</param>
        /// <returns>
        /// True if the suite contains a diagnostic with the specified name.
        /// </returns>
        /// 
	    public bool 
	    HasDiagnostic(String name)
	    {
		    return diagnostics.ContainsKey( name );
	    }

    }
}

//  ##########################################################################

File: java/Components/Strata.Diagnostic.Core/src/strata/diagnostic/core/common/DiagnosticSuite.java
// ##########################################################################
// # File Name:	DiagnosticSuite.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataInspector Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataInspector Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataInspector
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.diagnostic.core.common;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/**
 * Represents a suite of {@code IDiagnostic}s that are meant to be 
 * run as a whole or as part of a larger suite. {@code DiagnosticSuite} 
 * is a <a href="{@docRoot}/designpatterns/pat4c.htm">Composite</a> of 
 * {@code IDiagnostic}s.
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public 
class DiagnosticSuite 
	extends AbstractDiagnostic
{
	private Map<String,IDiagnostic> itsDiagnostics;
	
	/************************************************************************
	 * Creates a new DiagnosticSuite. 
	 *
	 * @param name	The name of the suite of diagnostics.
	 */
	public 
	DiagnosticSuite(String name)
	{
		super( name );
		itsDiagnostics = new LinkedHashMap<String,IDiagnostic>();
	}

	/************************************************************************
	 * {@inheritDoc}
	 * @return
	 */
	@Override
	public CompletionStage<IDiagnosticResult>
	runDiagnostic(IDiagnosticResult result)
	{
		return
			CompletableFuture.supplyAsync(
				() ->
				{
					try
					{
						List<CompletionStage<?>> results =
							new ArrayList<>();

						result.beginDiagnostic(this);
						beginDiagnosticMode();

						itsDiagnostics
							.values()
							.stream()
							.forEach(
								d -> results.add(d.runDiagnostic(result)));

						CompletableFuture
							.allOf(
								results.toArray(
									new CompletableFuture<?>[results.size()]))
							.get();
					}
					catch (DiagnosticAbortedException ae)
					{
						result.reportBeginFailure(this,ae);
					}
					catch (Exception e)
					{
						result.reportUnknownFailure(this,e);
					}
					finally
					{
						endDiagnosticMode();
						result.endDiagnostic(this);
					}

					return result;
				});
	}

	/************************************************************************
	 * Adds a IDiagnostic to the DiagnosticSuite. 
	 *
	 * @param 			d	The diagnostic being added.
	 * @precondition	this.hasDiagnostic( d.getName() ) == false
	 * @postcondition	this.hasDiagnostic( d.getName() ) == true
	 * @postcondition	this.getDiagnostic( d.getName() ) == d
	 */
	public void 
	addDiagnostic(IDiagnostic d)
	{
		itsDiagnostics.put( d.getName(),d );
	}
	
	/************************************************************************
	 * Removes the IDiagnostic with the specified 
	 * name from the DiagnosticSuite. 
	 *
	 * @param 			name	The name of the diagnostic being removed 
	 * 							from the suite.
	 * @precondition	this.hasDiagnostic( name ) == true
	 * @postcondition	this.hasDiagnostic( name ) == false
	 */
	public void 
	removeDiagnostic(String name)
	{
		itsDiagnostics.remove( name );
	}
	
	/************************************************************************
	 * Returns the IDiagnostic associated with the specified name. 
	 *
	 * @param 	name	The name of the diagnostic to be returned.
	 * @return	The diagnostic associated with name.
	 */
	public IDiagnostic 
	getDiagnostic(String name)
	{
		return itsDiagnostics.get( name );
	}

	/************************************************************************
	 * Returns all IDiagnostics contained in this suite.
	 *
	 * @return	All diagnostics contained in this suite.
	 */
	public Collection<IDiagnostic>
	getDiagnostics()
	{
		return itsDiagnostics.values();
	}

	/************************************************************************
	 * Queries the DiagnosticSuite if it contains a IDiagnostic 
	 * with the specified name. 
	 *
	 * @param 	name	The name of the diagnostic being queried.
	 * @return	True if the suite contains a diagnostic with the 
	 * 			specified name.
	 */
	public boolean 
	hasDiagnostic(String name)
	{
		return itsDiagnostics.containsKey( name );
	}
}

//##########################################################################
